overlays/nodejs/s3/s3Mixins.ts

```diff
@@ -0,0 +1,232 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as pulumi from "@pulumi/pulumi"
+import { Bucket } from "./bucket";
+import { BucketNotification } from "./bucketNotification";
+import * as lambda from "../lambda"
+
+/**
+ * Arguments to help customize a notification subscription for a bucket.
+ */
+export interface CommonBucketSubscriptionArgs {
+    /**
+     * An optional prefix or suffix to filter down notifications.  See
+     * aws.s3.BucketNotification.lambdaFunctions for more details.
+     */
+    filterPrefix?: string;
+    filterSuffix?: string;
+}
+
+export interface BucketEventSubscriptionArgs extends CommonBucketSubscriptionArgs {
+    /**
+     * Events to subscribe to. For example: "[s3:ObjectCreated:*]".  Cannot be empty.
+     */
+    events: string[];
+}
+
+/**
+ * Arguments to specifically control a subscription to 'ObjectCreated' notifications on a bucket.If
+ * more events than just 'ObjectCreated' events are desired, the 'subscribe' function should be used
+ * instead.
+ */
+export interface ObjectCreatedSubscriptionArgs extends CommonBucketSubscriptionArgs {
+    // The type of event to listen for.  '*' is the default if not provided.
+    event?: "*" | "Put" | "Post" | "Copy" | "CompleteMultipartUpload";
+}
+
+/**
+ * Arguments to specifically control a subscription to 'ObjectRemoved' notifications on a bucket. If
+ * more events than just 'ObjectRemoved' events are desired, the 'subscribe' function should be used
+ * instead.
+ */
+export interface ObjectRemovedSubscriptionArgs extends CommonBucketSubscriptionArgs {
+    // The type of event to listen for.  '*' is the default if not provided.
+    event?: "*" | "Delete" | "DeleteMarkerCreated";
+}
+
+// See https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html.
+export interface BucketEvent {
+    Records?: BucketRecord[];
+}
+
+export interface BucketRecord {
+    eventVersion: string;
+    eventSource: string;
+    awsRegion: string;
+    eventTime: string;
+    eventName: string;
+    userIdentity: {
+        principalId: string;
+    };
+    requestParameters: {
+        sourceIPAddress: string;
+    };
+    responseElements: {
+        "x-amz-request-id": string;
+        "x-amz-id-2": string;
+    };
+    s3: {
+        s3SchemaVersion: string;
+        configurationId: string;
+        bucket: {
+            name: string;
+            ownerIdentity: {
+                principalId: string;
+            },
+            arn: string;
+        };
+        object: {
+            key: string;
+            size: number;
+            eTag: string;
+            versionId?: string;
+            sequencer: string;
+        };
+    };
+}
+
+export type BucketEventHandler = lambda.EventHandler<BucketEvent, void>;
```

Comment:
By being a lambda.EventHandler, this is saying: you can either provide a JS function that we will serialize for you, or you can just provide the actual aws.lambda you want to be called.